### PR TITLE
Postgres cms port

### PR DESCRIPTION
**Pull request by ScottAP108**
- brought collection manager up to date to reflect some refinements made during hazy development

- integrated postgres tables into collection manager so that all logic will apply either to pg or mongo

- handling for geography range for post and patch requests

- massive file cleanup:  less and hbs folders gone, many other templates, etc. deleted

- package.json also slimmed down:  gulp, grunt, and handlebars gone

- removed nearly all images from static/img

### CODE REVIEW

File: client/js/models/__proto__.js [hunk 1]
```diff
@@ -37,16 +38,19 @@ module.exports = Object.assign( { }, require('../../../lib/Model'), require('eve
...
         } )
     },
 
     getCount() {
         return this.Xhr( { method: 'get', resource: this.resource, headers: this.headers || {}, qs: JSON.stringify( { countOnly: true } ) } )
-        .then( ( { result } ) => {
-            this.meta.count = result
-            return Promise.resolve( result )
+        .then( response => {
+            const count = Array.isArray( response ) ? response[0].count : response.result
```
Comment (cbaron) [inline client/js/models/__proto__.js]: Would be better if the backend responses were consistent.

File: client/js/models/__proto__.js [hunk 2]
```diff
@@ -55,7 +59,7 @@ module.exports = Object.assign( { }, require('../../../lib/Model'), require('eve
     moneyToReal( price ) { return parseFloat( price.replace( /\$|,/g, "" ) ) },
 
     patch( id, data ) {
-        return this.Xhr( { method: 'patch', id, resource: this.resource, headers: this.headers || {}, data: JSON.stringify( data || this.data ) } )
+        return this.Xhr( { method: 'patch', id, resource: this.resource, headers: Object.assign( { v2: true }, this.headers || {} ), data: JSON.stringify( data || this.data ) } )
```
Comment (cbaron) [inline client/js/models/__proto__.js]: What is still using v1?

File: client/js/views/CollectionManager.js [hunk 3]
```diff
@@ -43,26 +43,46 @@ module.exports = Object.assign( { }, require('./__proto__'), {
...
+                isPostgres = this.views.collections.collection.store.name[ collection ].isPostgres,
+                schema = this.views.collections.collection.store.name[ collection ].schema
+
             return {
                 insertion: { el: this.els.mainPanel },
-                model: Object.create( this.DocumentModel ).constructor( document, { resource: this.model.git('currentCollection') } ),
-                templateOpts: { message: `Delete "${document.label || document.name}" from ${this.model.git('currentCollection')}?` }
+                model: Object.create( this.DocumentModel ).constructor( document, {
+                    meta: { key: isPostgres ? 'id' : '_id' },
+                    resource: this.model.git('currentCollection')
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: `resource: collection`

Also, the client shouldn't care about whether the model is from postgres or mongo.  The client should just know to check for `id`, or `_id` for the key.

File: client/js/views/CollectionManager.js [hunk 4]
```diff
@@ -43,26 +43,46 @@ module.exports = Object.assign( { }, require('./__proto__'), {
...
+            const collection = this.model.git('currentCollection'),
+                meta = this.model.meta[ collection ] || { },
+                isPostgres = this.views.collections.collection.store.name[ collection ].isPostgres,
+                schema = this.views.collections.collection.store.name[ collection ].schema
+
             return {
-                model: Object.create( this.Model ).constructor( Object.assign( model, {
-                    collection: Object.create( this.DocumentModel ).constructor( [ ], { resource: this.model.git('currentCollection') } ),
-                    isDocumentList: true,
+                model: Object.create( this.Model ).constructor( Object.assign ( {
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: You don't need Object.assign here.

File: client/js/views/CollectionManager.js [hunk 5]
```diff
@@ -71,8 +91,15 @@ module.exports = Object.assign( { }, require('./__proto__'), {
                 disallowEnterKeySubmission: true,
                 insertion: { el: this.els.mainPanel },
                 model,
-                templateOpts: { heading: model.git('label') || model.git('name') },
-                Views: { }
+                templateOpts() {
+                    return Object.assign( { heading: model.heading }, model.meta.templateOptions )
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: Start using rest

```
templateOpts() {
    return { heading: model.heading, ...model.meta.templateOptions }
}
```

File: client/js/views/CollectionManager.js [hunk 6]
```diff
@@ -163,6 +191,8 @@ module.exports = Object.assign( { }, require('./__proto__'), {
 
         }
     },
+
+    cachedTables: [ 'deliveryoption', 'deliveryroute', 'groupdropoff', 'producefamily', 'seasonalAddOn', 'share', 'shareoption' ],
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: What's this for ?

File: client/js/views/CollectionManager.js [hunk 7]
```diff
@@ -174,39 +204,71 @@ module.exports = Object.assign( { }, require('./__proto__'), {
...
-
-        if( !collection ) return
+    createDocumentModel( data={} ) {
+        const collectionName = this.model.git('currentCollection'),
+            collection = this.views.collections.collection.store.name[ collectionName ],
+            meta = Object.assign( this.model.meta[ collectionName ] || { }, { noPlaceholder: true, key: collection.isPostgres ? 'id' : '_id' } )
 
-        if( type === 'documentList' ) return collection.clientData
+        let schema = this.model.git('currentCollection') === 'Pages'
+            ? collection.documents.find( doc => doc.name === data.label.replace( ' ', '' ) ).schema
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: What's this for ?

File: client/js/views/CollectionManager.js [hunk 8]
```diff
@@ -174,39 +204,71 @@ module.exports = Object.assign( { }, require('./__proto__'), {
...
         return Object.create( this.Model ).constructor(
             data,
-            Object.assign( { resource: this.model.git('currentCollection') }, schema )
+            Object.assign( {
+                meta,
+                resource: collectionName,
+                heading: this.getDisplayValue( meta, data, schema )
+            },
+                schema
+            )
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: { meta, resource: collectionName, heading: this.getDisplayValue( meta, data, schema ), ...schema }

File: client/js/views/CollectionManager.js [hunk 9]
```diff
@@ -174,39 +204,71 @@ module.exports = Object.assign( { }, require('./__proto__'), {
...
+            } ).join(' -- ')
+        } else {
+            value = datum.label || datum.name
+        }
+
+        return `<div><span>${value}</span></div>`
+    },
+
+    getDocument( collection, value ) {
+        const query = Number.isNaN( +value ) ? { name: value } : { id: value }
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: You shouldn't need to do this.  The model should know how to query itself.

File: client/js/views/CollectionManager.js [hunk 10]
```diff
@@ -254,7 +316,7 @@ module.exports = Object.assign( { }, require('./__proto__'), {
             const currentView = this.model.git('currentView'),
                 currentCollection = this.model.git('currentCollection'),
                 path = currentView === 'documentView'
-                    ? `/${currentCollection}/${this.views.documentView.model.git('name')}`
+                    ? `/${currentCollection}/${this.views.documentView.model.git('name') || this.views.documentView.model.git('id') || 'new-item'}`
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: Why did you add a feature to allow the back button to go to "new-item"?  Its not necessary and messy.

File: client/js/views/CollectionManager.js [hunk 11]
```diff
@@ -266,29 +328,38 @@ module.exports = Object.assign( { }, require('./__proto__'), {
             .catch( this.Error )
         } )
 
+        Promise.all( this.cachedTables.map( name => {
+            this[ name ] = Object.create( this.Model ).constructor( { }, { resource: name } )
+            return this[ name ].get( { storeBy: ['id'] } )
+        } ) )
+        .then( () => this.showProperView( true ) )
+        .catch( this.Error )
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: Are you handling cases where cached tables are updated ?

File: client/js/views/CollectionManager.js [hunk 12]
```diff
@@ -266,29 +328,38 @@ module.exports = Object.assign( { }, require('./__proto__'), {
...
     },
 
     showProperView() {
-        return (this.views.documentList ? Promise.resolve() : this.createDocumentList( this.model.git('currentCollection'), this.path.length === 2 ? false : true ) )
+        const path = this.path
+        return ( this.views.documentList ? Promise.resolve() : this.createDocumentList( this.model.git('currentCollection'), this.path.length === 2 ? false : true ) )
         .then( () =>
-            this.path.length === 2
-                ? this.getDocument()
+            path.length === 2 && path[1] !== 'new-item'
```
Comment (cbaron) [inline client/js/views/CollectionManager.js]: I don't think its necessary to handle `new-item` in the routing, it just over complicates things for little value.

Comment (cbaron) [reply client/js/views/CollectionManager.js]: If the user is creating, then clicks something else, then hits back, that's an edge case and they can click create again.

File: client/js/views/List.js [hunk 13]
```diff
@@ -130,7 +130,9 @@ module.exports = Object.assign( { }, Super, {
         this.fetching = true
         if( nextPage ) this.model.set( 'skip', this.model.git('skip') + this.model.git('pageSize') )
 
-        return this.collection.get( { query: { skip: this.model.git('skip'), limit: this.model.git('pageSize'), sort: this.model.git('sort') } } )
+        const query = this.model.git('isPostgres') ? { } : { skip: this.model.git('skip'), limit: this.model.git('pageSize'), sort: this.model.git('sort') }
```
Comment (cbaron) [inline client/js/views/List.js]: Don't you think the postgres stuff should be paginated too ?

File: client/js/views/templates/FieldError.js [hunk 14]
```diff
@@ -0,0 +1,2 @@
+module.exports = p =>
```
Comment (cbaron) [inline client/js/views/templates/FieldError.js]: I think we need more simple, reusable templates used throughout the app.

File: dal/postgres.js [hunk 15]
```diff
@@ -22,6 +22,8 @@ module.exports = Object.create( Object.assign( {}, MyObject, {
         }
     },
 
+    excludedCmsTables: [ 'csaTransaction', 'role', 'spatial_ref_sys', 'tablemeta', 'transaction' ],
```
Comment (cbaron) [inline dal/postgres.js]: I don't think this logic should be in the dal.  Postgres doesn't care about the CMS.

File: dal/postgres.js [hunk 16]
```diff
@@ -106,6 +108,34 @@ module.exports = Object.create( Object.assign( {}, MyObject, {
                 )
             )
         )
+        .then( () => {
+            return this.query( "SELECT * FROM tablemeta" )
```
Comment (cbaron) [inline dal/postgres.js]: Stop using tablemeta.

File: dal/postgres.js [hunk 17]
```diff
@@ -106,6 +108,34 @@ module.exports = Object.create( Object.assign( {}, MyObject, {
...
+                            }
+                        } )
+                    }
+
+                    return { name: row.name, label: row.label, schema, isPostgres: true }
+                } )
+
+                return Promise.resolve()
+            } )
+        } )
```
Comment (cbaron) [inline dal/postgres.js]: I do not like this at all.  The `getTableData` method already does everything for you.  This transform can be done in another call or on the client. ( Just like how its done in Mongo )

File: models/About.js [hunk 18]
```diff
@@ -1,67 +1,60 @@
...
-        ] )
-    }
-
+    attributes: [
+        {
+            name: 'sectionOne',
+            label: 'Section 1',
+            range: [
+                {
+                    name: 'heading',
```
Comment (cbaron) [inline models/About.js]: If the `label` is just a capitalized first letter of name, you can omit, and have a transform in the client.

File: resources/Collection.js [hunk 19]
```diff
@@ -26,15 +26,15 @@ Object.assign( Collection.prototype, Base.prototype, {
...
                                   name, 
-                                  schema: this.Mongo.model[name].schema,
-                                  clientData: this.Mongo.model[ name ].clientData
+                                  schema: this.Mongo.model[name]
                               } ) )
                           } )
-                        : ( { name, schema: this.Mongo.model[name].schema, clientData: this.Mongo.model[ name ].clientData } ) 
+                        : ( { name, schema: this.Mongo.model[name] } )
                 )
+                .concat( this.Postgres.cmsModels )
```
Comment (cbaron) [inline resources/Collection.js]: Do the postgres cms stuff here, not in dal/Postgres

File: client/js/models/__proto__.js (hunk 1 above)
Comment (ScottAP108) [reply client/js/models/__proto__.js]: I'd have to change the way either mongo or postgres responds, which would cause problems in many other places.

File: client/js/models/__proto__.js (hunk 2 above)
Comment (ScottAP108) [reply client/js/models/__proto__.js]: Most normal requests, I believe. I should've bitten the bullet a long time ago and just updated the entire server to newer specs. It has caused me much grief.

File: client/js/views/CollectionManager.js (hunk 3 above)
Comment (ScottAP108) [reply client/js/views/CollectionManager.js]: The problem is that the client needs to know whether it's a postgres table before the data is even fetched, so you can't just check for _id or id because you don't have data. You have to have some way to tell, and I felt my way was the simplest. You could hardcode that info into the collection manager model for each table, but I don't like that.

File: client/js/views/CollectionManager.js (hunk 5 above)
Comment (ScottAP108) [reply client/js/views/CollectionManager.js]: I've been avoiding rest/spread in PW for the sake of consistency. I think it should be used everywhere or not at all, and I'm going to start using it everywhere in Maricann.

File: client/js/views/CollectionManager.js (hunk 6 above)
Comment (ScottAP108) [reply client/js/views/CollectionManager.js]: These are tables that are referenced in foreign keys. I store them all upfront in postRender so that I can synchronously handle fk lookups from that point on. It's a negligible upfront cost and saves me from awkwardly injecting a bunch of requests into the rendering process.

File: client/js/views/CollectionManager.js (hunk 7 above)
Comment (ScottAP108) [reply client/js/views/CollectionManager.js]: It's how I handle the models for the main site pages. `Pages` is itself a collection, but unlike other collections, its documents also need a server model. See `resources/Collection` for clarification.

File: client/js/views/CollectionManager.js (hunk 3 above)
Comment (cbaron) [reply client/js/views/CollectionManager.js]: You don't have to change it, but I think the server response should include the `key` for a table/collection.  The client should definitely not be in the business of knowing what postgres or mongo is.

File: client/js/views/CollectionManager.js (hunk 5 above)
Comment (cbaron) [reply client/js/views/CollectionManager.js]: I disagree.  We need to incrementally improve things bit by bit when we come across them.  This way we can avoid 180 file changes in a PR.

File: client/js/views/CollectionManager.js (hunk 10 above)
Comment (ScottAP108) [reply client/js/views/CollectionManager.js]: I don't think you're understanding this. Before, when you clicked the Add button to create a new document, the code would stick `undefined` at the end of the url. I just did this instead. It has nothing to do with the back button. You literally cannot go to this route with the back button as there's nothing beyond it. It's only there when creating a new document and disappears once you make the document or go back to the document list.

File: client/js/views/CollectionManager.js (hunk 7 above)
Comment (cbaron) [reply client/js/views/CollectionManager.js]: Matching like

```
doc.name === data.label.replace( ' ', '' )
```

is not good.

File: client/js/views/CollectionManager.js (hunk 11 above)
Comment (ScottAP108) [reply client/js/views/CollectionManager.js]: No. That's a good point. I should do that.

File: client/js/views/List.js (hunk 13 above)
Comment (ScottAP108) [reply client/js/views/List.js]: All the longest tables had to do with members, and those are no longer here. I think `produce` has the most items of the tables here, which is not much. I prefer not having pagination so they can just do a ctrl+f search to find what they want. If you introduce pagination, you should also have a search bar, I think, in case the desired document is on another page.

File: dal/postgres.js (hunk 16 above)
Comment (ScottAP108) [reply dal/postgres.js]: Tablemeta is the only source for a label for the postgres tables, so I need it. Otherwise the collection names in the sidebar will all be a mess.

File: dal/postgres.js (hunk 17 above)
Comment (ScottAP108) [reply dal/postgres.js]: I can move it to `resources/Collection` so that the code only runs for users of the collection manager, but I see nothing wrong with the code itself. I am simply gathering all of the information I need to use later on for each table.

File: client/js/views/CollectionManager.js (hunk 3 above)
Comment (ScottAP108) [reply client/js/views/CollectionManager.js]: Ok that makes sense.

File: client/js/views/CollectionManager.js (hunk 10 above)
Comment (ScottAP108) [reply client/js/views/CollectionManager.js]: You just get sent back to the document list.
